contracts/token/ERC20Rebasable.sol

```diff
@@ -0,0 +1,358 @@
+
+    /// @notice Amount of shares (locked wstETH amount) owned by the holder.
+    function _getShares() internal pure returns (mapping(address => uint256) storage) {
+        return SHARES_POSITION.storageMapAddressAddressUint256();
+    }
+
+    /// @notice The total amount of shares in existence.
+    function _getTotalShares() internal view returns (uint256) {
+        return TOTAL_SHARES_POSITION.getStorageUint256();
+    }
+
+    /// @notice Set total amount of shares.
+    function _setTotalShares(uint256 _newTotalShares) internal {
+        TOTAL_SHARES_POSITION.setStorageUint256(_newTotalShares);
+    }
+
+    /// @dev Moves amount_ of tokens from sender_ to recipient_
+    /// @param from_ An address of the sender of the tokens
+    /// @param to_  An address of the recipient of the tokens
+    /// @param amount_ An amount of tokens to transfer
+    function _transfer(
+        address from_,
+        address to_,
+        uint256 amount_
+    ) internal onlyNonZeroAccount(from_) onlyNonZeroAccount(to_) {
+        uint256 sharesToTransfer = _getSharesByTokens(amount_);
+        _transferShares(from_, to_, sharesToTransfer);
+        _emitTransferEvents(from_, to_, amount_ ,sharesToTransfer);
+    }
+
+    /// @dev Updates owner_'s allowance for spender_ based on spent amount_. Does not update
+    ///     the allowance amount in case of infinite allowance
+    /// @param owner_ An address of the account to spend allowance
+    /// @param spender_ An address of the spender of the tokens
+    /// @param amount_ An amount of allowance spend
+    function _spendAllowance(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal {
+        uint256 currentAllowance = _getTokenAllowance()[owner_][spender_];
+        if (currentAllowance == type(uint256).max) {
+            return;
+        }
+        if (amount_ > currentAllowance) {
+            revert ErrorNotEnoughAllowance();
+        }
+        unchecked {
+            _approve(owner_, spender_, currentAllowance - amount_);
+        }
+    }
+
+    /// @dev Sets amount_ as the allowance of spender_ over the owner_'s tokens
+    /// @param owner_ An address of the account to set allowance
+    /// @param spender_ An address of the tokens spender
+    /// @param amount_ An amount of tokens to allow to spend
+    function _approve(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal virtual onlyNonZeroAccount(owner_) onlyNonZeroAccount(spender_) {
+        _getTokenAllowance()[owner_][spender_] = amount_;
+        emit Approval(owner_, spender_, amount_);
+    }
+
+    function _sharesOf(address account_) internal view returns (uint256) {
+        return _getShares()[account_];
+    }
+
+    function _getTokensByShares(uint256 sharesAmount_) internal view returns (uint256) {
+        (uint256 tokensRate, uint256 decimals) = _getTokenRateAndDecimal();
+        return (sharesAmount_ * tokensRate) / (10 ** decimals);
```

Comment:
e2e test is needed to check that `decimals` dimension stays correct starting from L1 side up to operations with stETH on L2

i.e.,
- you wrap 1.273 stETH on L1 to X wstETH
- you wrap 1.273 stETH on L2 to Y wstETH

⇒ invariant: X == Y

---
the same is for unwrap